# SLR - MiniBatchGD (FAKE DATASET)

In [ ]:
import numpy as np

X_simple = 2 * np.random.rand(100, 1)
y_simple = 4 + 3 * X_simple + np.random.randn(100, 1)

X_simple_b = np.c_[np.ones((100, 1)), X_simple]

n_iterations = 500
minibatch_size = 20
learning_rate = 0.01

theta_simple = np.random.randn(2, 1)
for iteration in range(n_iterations):
    shuffled_indices = np.random.permutation(100)
    X_simple_b_shuffled = X_simple_b[shuffled_indices]
    y_simple_shuffled = y_simple[shuffled_indices]
    for i in range(0, 100, minibatch_size):
        xi = X_simple_b_shuffled[i:i+minibatch_size]
        yi = y_simple_shuffled[i:i+minibatch_size]
        gradients = 2/minibatch_size * xi.T.dot(xi.dot(theta_simple) - yi)
        theta_simple = theta_simple - learning_rate * gradients

print("Simple Linear Regression Weights (theta):")
print(theta_simple)

Simple Linear Regression Weights (theta):
[[3.60091033]
 [3.28858527]]


# MLR - MiniBatchGD (FAKE DATASET)

In [ ]:
import numpy as np

X_multiple = 2 * np.random.rand(100, 3)
y_multiple = 4 + 3 * X_multiple[:, 0] + 5 * X_multiple[:, 1] + 2 * X_multiple[:, 2] + np.random.randn(100)
y_multiple = y_multiple.reshape(-1, 1)

X_multiple_b = np.c_[np.ones((100, 1)), X_multiple]

n_iterations = 500
minibatch_size = 20
learning_rate = 0.01

theta_multiple = np.random.randn(4, 1)

for iteration in range(n_iterations):
    shuffled_indices = np.random.permutation(100)
    X_multiple_b_shuffled = X_multiple_b[shuffled_indices]
    y_multiple_shuffled = y_multiple[shuffled_indices]
    for i in range(0, 100, minibatch_size):
        xi = X_multiple_b_shuffled[i:i+minibatch_size]
        yi = y_multiple_shuffled[i:i+minibatch_size]
        gradients = 2/minibatch_size * xi.T.dot(xi.dot(theta_multiple) - yi)
        theta_multiple = theta_multiple - learning_rate * gradients

print("Multiple Linear Regression Weights (theta):")
print(theta_multiple)

Multiple Linear Regression Weights (theta):
[[3.93967946]
 [2.85560309]
 [5.17028243]
 [1.95198939]]


# SLR - MiniBatchGD (KAGGLE SAMPLE DATASET)

In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv('/content/sample_data/Salary_dataset.csv')
X_simple = df['YearsExperience'].values.reshape(-1, 1)
y_simple = df['Salary'].values.reshape(-1, 1)

X_simple_b = np.c_[np.ones((X_simple.shape[0], 1)), X_simple]

n_iterations = 500
minibatch_size = 20
learning_rate = 0.01

theta_simple = np.random.randn(2, 1)

for iteration in range(n_iterations):
    shuffled_indices = np.random.permutation(X_simple.shape[0])
    X_simple_b_shuffled = X_simple_b[shuffled_indices]
    y_simple_shuffled = y_simple[shuffled_indices]
    for i in range(0, X_simple.shape[0], minibatch_size):
        xi = X_simple_b_shuffled[i:i+minibatch_size]
        yi = y_simple_shuffled[i:i+minibatch_size]
        gradients = 2/minibatch_size * xi.T.dot(xi.dot(theta_simple) - yi)
        theta_simple = theta_simple - learning_rate * gradients

print("Simple Linear Regression Weights (theta):")
print(theta_simple)

Simple Linear Regression Weights (theta):
[[23817.39932834]
 [ 9552.96718135]]



## **Comparison: Simple Linear Regression vs Ridge Regularized SLR (MiniBatchGD)**

| Model                             | Intercept (θ₀) | Slope (θ₁) |
| --------------------------------- | -------------- | ---------- |
| **Simple Linear Regression**      | 23817.40       | 9552.97    |
| **Ridge Regularized SLR (α=0.1)** | 15728.96       | 10659.09   |

###**Observations**

* **Intercept (θ₀):**

  * Ridge regularization significantly **reduced the intercept**, shifting the baseline downward.
* **Slope (θ₁):**

  * Ridge slightly **increased the slope**, emphasizing the relationship between experience and salary while controlling coefficient magnitude.
* **Effect of Ridge Regularization:**

  * Helps **prevent overfitting** by penalizing large weights.
  * Balances bias-variance trade-off by shrinking coefficients toward zero.

###**Conclusion**

* **Simple Linear Regression:** Fits data directly; may lead to overfitting with noisy data or small batches.
* **Ridge Regularized SLR:** Reduces large coefficients and adjusts the intercept, producing a more **generalized and stable model** for prediction.



# MLR - MiniBatchGD (KAGGLE SAMPLE DATASET)

In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv('/content/sample_data/Student_Performance.csv')
X_multiple = df[['Hours Studied', 'Previous Scores', 'Sleep Hours', 'Sample Question Papers Practiced']].values
y_multiple = df['Performance Index'].values
y_multiple = y_multiple.reshape(-1, 1)

X_multiple_b = np.c_[np.ones((df.shape[0], 1)), X_multiple]

n_iterations = 500
minibatch_size = 20
learning_rate = 0.0001

theta_multiple = np.random.randn(X_multiple_b.shape[1], 1)

for iteration in range(n_iterations):
    shuffled_indices = np.random.permutation(df.shape[0])
    X_multiple_b_shuffled = X_multiple_b[shuffled_indices]
    y_multiple_shuffled = y_multiple[shuffled_indices]
    for i in range(0, df.shape[0], minibatch_size):
        xi = X_multiple_b_shuffled[i:i+minibatch_size]
        yi = y_multiple_shuffled[i:i+minibatch_size]
        gradients = 2/minibatch_size * xi.T.dot(xi.dot(theta_multiple) - yi)
        theta_multiple = theta_multiple - learning_rate * gradients

print("Multiple Linear Regression Weights (theta):")
print(theta_multiple)

Multiple Linear Regression Weights (theta):
[[-24.98696361]
 [  2.6798129 ]
 [  0.96099039]
 [ -0.04813742]
 [  0.07537264]]


## **Comparison: Multiple Linear Regression vs Ridge Regularized MLR (MiniBatchGD)**

| Model                             | θ₀ (Intercept) | θ₁ (Hours Studied) | θ₂ (Previous Scores) | θ₃ (Sleep Hours) | θ₄ (Sample Papers) |
| --------------------------------- | -------------- | ------------------ | -------------------- | ---------------- | ------------------ |
| **Multiple Linear Regression**    | -24.987        | 2.6798             | 0.9610               | -0.0481          | 0.0754             |
| **Ridge Regularized MLR (α=0.1)** | -7.053         | 2.2876             | 0.8402               | -1.0768          | -0.1765            |

###**Observations**

* **Intercept (θ₀):** Ridge regularization **shifted the intercept upward**, closer to zero.
* **Coefficients (θ₁–θ₄):**

  * Ridge shrinks coefficients slightly to **prevent overfitting**.
  * Negative and positive contributions are moderated; small or noisy effects are reduced (e.g., θ₄ decreased).
* **Effect of Ridge:** Helps maintain **generalization** by penalizing large weights.

###**Conclusion**

* **Standard MiniBatchGD MLR:** Fits dataset directly, may be sensitive to noise in small batches.
* **Ridge Regularized MiniBatchGD MLR:** Reduces coefficient magnitudes, balances feature contributions, and improves model stability.

